Python/chapter01/1.7 - Rotate Matrix/miguel_1.7_sol.py

```diff
@@ -0,0 +1,174 @@
+"""
+Python version 3.7.0
+1.7 - Rotate Matrix
+Given an image represented by an NxN matrix, where each pixel in the image is 4 bytes, write a method to rotate
+the image by 90 degrees.  Can you do this in place?
+"""
+import unittest
+from typing import List
+
+
+def rotate_matrix(matrix: List[List[int]], N: int) -> List[List[int]]:
+    """
+    Rotate matrix will rotate the given matrix by 90 degrees.
+    Runtime: O(N^2), asymptotic runtime depends on N
+    Space Complexity: O(N^2), creating a new matrix of NxN called 'rotated'
+    :param matrix: an NxN matrix
+    :param N: the size of the matrix (NxN)
+    :return: a newly rotated matrix
+    """
+    rotated = [[0 for i in range(N)] for j in range(N)]
+    for i in range(N):
+        for j in range(N):
+            rotated[j][(N-1)-i] = matrix[i][j]
+    return rotated
+
+
+def rotate_matrix_in_place(matrix: List[List[int]], start_row: int, start_col: int, N: int) -> List[List[int]]:
+    """
+    Does the same as rotate_matrix, but in place.
+    Runtime: O(N^2), asymptotic runtime depends on N. We make N^2 swaps.
+    Space Complexity: O(1), constant amount of temp variables that does not depend on N.
+                        Note: I am assuming that compiler will optimize the tail recursion.
```

Comment:
this assumption is unfortunately false and will most likely never be true: http://neopythonic.blogspot.com/2009/04/tail-recursion-elimination.html

Reply:
Does that mean that the space complexity will not be O(1)?

Reply:
correct: Python will always use O(n) stack space.